### PR TITLE
Updated ref matchers to use docpipe matchers, to avoid duplicating code

### PR DESCRIPTION
**Pull request by longhotsummer**
* add a mixin to make docpipe matchers work with Document objects
* apply mixin to existing docpipe matchers
* implement additional matchers on top of docpipe
* the old 'refs' matcher which handled both "Act X of YYYY" and "the Constitution" is now split into two
* the new alias matcher handles "the Constitution" and friends
* also adds a French act number matcher

## outstanding

* migrate refs-subtype to use new mechanism

### CODE REVIEW

File: indigo/analysis/refs/works.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+class CommonTitlesCitationMatcher(DocumentPatternMatcherMixin, CitationMatcher):
+    """Finds references to works based "the XXX Act" patterns. """
+    # only supports English
+    locale = (None, 'eng', None)
+
+    html_candidate_xpath = ".//text()[contains(., ' the ') and contains(., ' Act') and not(ancestor::a)]"
+    xml_candidate_path = ".//text()[contains(., ' the ') and contains(., ' Act') and not(ancestor::ns:ref)]"
+
+    pattern_re = re.compile(r'\b(?P<pre>the\s+(?P<title>((?!the)\S+\s+){1,7})Act)\s*,?(\s*of\s+)?\s*(?P<year>[12][0-9]{3})?\b')
```
Comment (goose-life) [inline indigo/analysis/refs/works.py:21]: am I reading this right that "XXX" in 'the XXX Act' can only be 7 characters long and contain no spaces?

File: indigo/analysis/refs/works.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    # map from titles to FRBR URIs
+    titles = None
+
+    def setup(self, frbr_uri, text=None, root=None):
+        self.setup_titles(frbr_uri)
+        super().setup(frbr_uri, text, root)
+
+    def setup_titles(self, frbr_uri):
+        self.titles = self.titles.get(frbr_uri.place) or self.titles.get(frbr_uri.country) or {}
```
Comment (goose-life) [inline indigo/analysis/refs/works.py:31]: but where do they actually get set up lol 😆

File: indigo/analysis/refs/works.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+            qs = self.get_queryset(frbr_uri, country, locality)
+
+            # a work can't cite something published in the future
+            date = frbr_uri.expression_date
+            if date and date.startswith('@'):
+                qs = qs.filter(date__lte=date[1:5])
+
+            for we in qs:
+                self.titles[we["title"]] = we["frbr_uri"]
```
Comment (goose-life) [inline indigo/analysis/refs/works.py:80]:
```suggestion
            for w in qs:
                self.titles[w["title"]] = w["frbr_uri"]
```

File: indigo/analysis/refs/works.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    def get_queryset(self, frbr_uri, country, locality):
+        # load titles for WorkExpressions in this country and/or locality
```
Comment (goose-life) [inline indigo/analysis/refs/works.py:83]:
```suggestion
        # load titles for Works in this country and/or locality
```

File: indigo_za/refs.py [hunk 5]
```diff
@@ -1,104 +0,0 @@
...
-            group = 1
-
-        ref = etree.Element(self.marker_tag)
-        ref.text = match.group(group)
-        ref.set('href', self.make_href(match))
-        return ref, match.start(group), match.end(group)
-
-
-@plugins.register('refs')
-class RefsFinderAFRza(RefsFinderENGza):
```
Comment (goose-life) [inline indigo_za/refs.py:68]: will we need to add a new afr plugin?

**Review by goose-life — APPROVED** [review]
couple of minor queries and suggestions

File: indigo/analysis/refs/works.py (hunk 1 above)
Comment (longhotsummer) [reply indigo/analysis/refs/works.py:21]: no, it can be up to 7 words long `(\S+\s+){1,7}`

File: indigo_za/refs.py (hunk 5 above)
Comment (longhotsummer) [reply indigo_za/refs.py:68]: oh, yes, similar to how it is done for FRA

File: indigo/analysis/refs/works.py (hunk 1 above)
Comment (goose-life) [reply indigo/analysis/refs/works.py:21]: the makes a lot more sense :)

File: indigo/analysis/refs/works.py (hunk 2 above)
Comment (longhotsummer) [reply indigo/analysis/refs/works.py:31]: In `WorkTitlesCitationMatcher`

File: indigo_za/refs.py (hunk 5 above)
Comment (longhotsummer) [reply indigo_za/refs.py:68]: added

**Review by goose-life — APPROVED** [review]

**Pull request merged**